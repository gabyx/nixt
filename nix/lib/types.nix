{
  inputs,
  cell,
}: let
  inherit (inputs) nixpkgs std;
  lib = nixpkgs.lib // builtins;

  inherit (std.inputs.yants) string bool path attrs list struct defun either option;
in rec {
  Schema = struct "Schema" {
    __schema = string;
    settings = Settings;
    testSpec = list Block;
  };

  Settings = struct "Settings" {
    list = bool;
    watch = bool;
    verbose = bool;
    trace = bool;
  };

  Block = struct "Block" {
    path = path;
    suites = list TestSuite;
  };

  TestSuite = struct "TestSuite" {
    name = string;
    cases = list TestCase;
  };

  TestCase = struct "TestCase" {
    name = string;
    expressions = list bool;
  };
}
