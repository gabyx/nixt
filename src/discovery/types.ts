export class TestCase {
  name: string;
  result?: boolean;
  error?: string;

  constructor(name: string) {
    this.name = name;
  }
}

export class TestSuite {
  name: string;
  cases: Record<string, TestCase>;

  constructor(name: string) {
    this.name = name;
    this.cases = {};
  }
}

export class TestFile {
  path: string;
  suites: Record<string, TestSuite>;
  importError?: string;

  constructor(path: string) {
    this.path = path;
    this.suites = {};
  }
}