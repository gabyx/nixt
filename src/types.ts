import { z } from "zod";

export const testCase = z.object({
  name: z.string(),
  expressions: z.boolean().array(),
  error: z.string().optional(),
});

export type TestCase = z.infer<typeof testCase>;

export const testSuite = z.object({
  name: z.string(),
  cases: testCase.array(),
});

export type TestSuite = z.infer<typeof testSuite>;

export const testFile = z.object({
  path: z.string(),
  suites: testSuite.array(),
  importError: z.string().optional(),
});

export type TestFile = z.infer<typeof testFile>;

export const schemaVer = "v0.0";
export const schema = z
  .object({
    __schema: z.string(),
    settings: z.object({
      list: z.boolean(),
      watch: z.boolean(),
      verbose: z.boolean(),
      trace: z.boolean(),
    }),
    testSpec: testFile.array(),
  })
  .required();

export type Schema = z.infer<typeof schema>;

export type CliArgs = {
  paths: string[];
  watch: boolean;
  verbose: boolean;
  showTrace: boolean;
  list: boolean;
  recurse: boolean;
  debug: boolean;
};

export type TOptions = {
  watch: boolean;
  verbose: boolean;
  showTrace: boolean;
  list: boolean;
  recurse: boolean;
  debug: boolean;
};
