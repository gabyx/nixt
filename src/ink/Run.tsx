import React, { FC } from "react";

import { CliArgs } from "../types.js";
import { Text } from "ink";

export const Run: FC<CliArgs> = (args) => {
  return <Text>{JSON.stringify(args, null, 4)}</Text>;
};
