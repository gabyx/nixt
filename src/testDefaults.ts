import { schemaVer } from "./types.js";

export const defaultArgs = {
  paths: [],
  watch: false,
  verbose: false,
  showTrace: false,
  list: false,
  recurse: true,
  debug: false,
};

export const defaultRegistry = {
  __schema: schemaVer,
  settings: {
    list: false,
    watch: false,
    verbose: false,
    trace: false,
  },
  testSpec: [
    {
      path: "./dummy.nix",
      suites: [
        {
          name: "Dummy",
          cases: [
            {
              name: "is a dummy suite",
              expressions: [true],
            },
          ],
        },
      ],
    },
  ],
};
