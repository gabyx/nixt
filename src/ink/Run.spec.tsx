import React from "react";

import { render } from "ink-testing-library";
import { defaultArgs } from "../testDefaults.js";
import { CliArgs } from "../types.js";
import { Run } from "./Run.js";

describe("Ink Run", () => {
  const Sut = Run;
  let args: CliArgs;

  beforeEach(() => {
    args = { ...defaultArgs };
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("is defined", () => {
    expect(Sut).toBeDefined();
  });

  it("renders expected default frame", () => {
    const sut = render(<Sut {...args} />);

    expect(sut.lastFrame()).toMatchSnapshot();
  });

  it.todo("runs in flake mode when no path is given", () => {
    args.paths = [];

    render(<Sut {...args} />);
  });

  it.todo("runs in standalone mode when a path is given", () => {
    args.paths = ["."];

    render(<Sut {...args} />);
  });

  it.todo("throws when no path is given and the registry in invalid", () => {
    args.paths = [];
    // TODO: Mock registry

    render(<Sut {...args} />);
  });

  it.todo("calls renderService once", () => {
    render(<Sut {...args} />);
  });
});
