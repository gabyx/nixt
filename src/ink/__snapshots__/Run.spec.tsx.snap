// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ink Run > renders expected default frame 1`] = `
"{
    \\"paths\\": [],
    \\"watch\\": false,
    \\"verbose\\": false,
    \\"showTrace\\": false,
    \\"list\\": false,
    \\"recurse\\": true,
    \\"debug\\": false
}"
`;
